PR: https://github.com/odoo/odoo/pull/257280

From: 4588cc25d232871530f992a85aab0b41c642bab1
From: Saif (segn)
Date: 2026-04-20 14:47:13

Breaking data model changes scores: del:1 + add:0, change matches:
-    foldable = fields.Boolean(string="Foldable", help="By default, we always unfold the lines that can be. If this is checked, the line won't be unfolded by default, and a folding button will be displayed.")

Total Changes: 31

[IMP] account: foldability as selection field

* After the groupby on the report^1 was added and support of groupby on
aggregation being added, we don't want to have the groupby being applied
by default on the aggregation lines otherwise the report will be
unreadable.
Therefore, changing the foldable boolean to a computed stored selection
field, allowing for default behaviour to be as intended while enabling
the user to change them on the line if needed.

* In upgrade_code: removing the unnecessary foldable field in reports data files if they
will be correctly computed.

* For inconsistencies reported by script, manually set the foldability values
on the reports.

^1 : https://github.com/odoo/odoo/commit/05f1d84347c294f35b239d024e3b3c82bb6a9fb6

task-5891053

Part-of: odoo/odoo#257280
Related: odoo/enterprise#112849
Related: odoo/documentation#17463
Related: odoo/upgrade#9835
Signed-off-by: Olivier Colson (oco) <oco@odoo.com>

================================= pseudo patch: =================================

--- a/addons/account/models/account_report.py
+++ b/addons/account/models/account_report.py
@@ -407,7 +407,16 @@ class AccountReportLine(models.Model):
     )
     sequence = fields.Integer(string="Sequence")
     code = fields.Char(string="Code", help="Unique identifier for this line.")
-    foldable = fields.Boolean(string="Foldable", help="By default, we always unfold the lines that can be. If this is checked, the line won't be unfolded by default, and a folding button will be displayed.")
+    foldability = fields.Selection(
+        selection=[
+            ('always_unfolded', 'Always Unfolded'),
+            ('never_unfolded', 'Never Unfolded'),
+            ('foldable', 'Foldable'),
+        ],
+        compute='_compute_foldability',
+        store=True,
+        readonly=False,
+    )
     print_on_new_page = fields.Boolean('Print On New Page', help='When checked this line and everything after it will be printed on a new page.')
     action_id = fields.Many2one(string="Action", comodel_name='ir.actions.actions', help="Setting this field will turn the line into a link, executing the action when clicked.")
     hide_if_zero = fields.Boolean(string="Hide if Zero", help="This line and its children will be hidden when all of their columns are 0.")
@@ -444,6 +453,21 @@ class AccountReportLine(models.Model):
             if report_line.parent_id:
                 report_line.horizontal_split_side = report_line.parent_id.horizontal_split_side
 
+    @api.depends('create_date')
+    def _compute_foldability(self):
+        for line in self:
+            if line.foldability:
+                continue
+            has_groupby = bool(line.groupby or line.user_groupby or line.report_id.groupby or line.report_id.user_groupby)
+            if line.children_ids:
+                line.foldability = 'always_unfolded'
+            elif has_groupby and all(expr.engine not in ('aggregation', 'external') for expr in line.expression_ids):
+                line.foldability = 'foldable'
+            elif any(expr.engine in ('aggregation', 'external') for expr in line.expression_ids):
+                line.foldability = 'never_unfolded'
+            else:
+                line.foldability = 'always_unfolded'
+
     @api.depends('groupby', 'expression_ids.engine')
     def _compute_user_groupby(self):
         for report_line in self:
@@ -454,6 +478,11 @@ class AccountReportLine(models.Model):
             except UserError:
                 report_line.user_groupby = report_line.groupby
 
+    @api.onchange('user_groupby')
+    def _onchange_user_groupby(self):
+        if self.user_groupby and self.foldability == 'never_unfolded':
+            self.foldability = 'foldable'
+
     @api.constrains('parent_id')
     def _validate_groupby_no_child(self):
         for report_line in self:
